### PR TITLE
run bare rust CLI binary, no node

### PR DESCRIPTION
**Pull request by alexeagle**
BREAKING CHANGES:

- the term "transpile" is not typically used in swc, so the underlying rule is renamed `swc_compile` matching the command we run `swc compile ...`
- We force users to upgrade to newest SWC version that is compatible. We print a nice error if your swc is older than 1.3.25

### CODE REVIEW

**Comment by alexeagle** [comment]
Okay this has been hard and I haven't found much focus time for it.

As far as I know, https://github.com/swc-project/swc/issues/5171 was the last bug preventing Bazel from using the pure-rust CLI.

We will have some rules_swc API changes:
- [ ] just mirror the rust CLI, no npm packages
- [ ] allow `integrity_hashes={"darwin-arm64": "sha512-DuB...` to be passed along with version number in WORKSPACE so users don't have to rely on our mirroring
- [ ] look for .swcrc in the package and use it if present, like we do for tsconfig.json in ts_project
- [ ] maybe the swcrc file becomes required to make the rust CLI work. Accept a dictionary to dynamically generate it, just like we do with tsconfig.json
- [ ] get all the tests passing again, and document the breaking changes that were required

**Comment by alexeagle** [comment]
Update: it's still necessary to provide a fake tty to make the swcx binary accept a folder, now trying to make a `tty_binary` rule in https://github.com/aspect-build/bazel-lib/pull/251

This keeps getting stuck because it's a side-project for me. We need some sponsorship for this work. Filed #88 so that the PR has an associated issue.

**Comment by realtimetodie** [comment]
I'm working on this

**Comment by realtimetodie** [comment]
Fixed an upstream bug, see https://github.com/aspect-build/rules_swc/issues/88

**Comment by alexeagle** [comment]
awesome, LMK if you'd like some help getting this rebased or anything

**Comment by realtimetodie** [comment]
Yes, I already made a rebase in https://github.com/realtimetodie/rules_swc/tree/swc_rust and got `bazel build //... ` running with a custom swc cli and the changes from https://github.com/swc-project/swc/pull/6708. In addition, I found new bugs in the swc cli and will create subsequent PRs upstream.

`bazel test //...` 13 tests pass and 2 fail locally. All failed tests are related to the `--output-dir` option of the swc cli compile command. It is due to the cwd. The cwd is set to the execroot directory and the swc cli can not find the auto-generated mock test files that are written into the bazel-out directory. But I think this might be related to my custom binary location, see

https://github.com/realtimetodie/rules_swc/commit/ac516ac2256f9a273c862a966d8485a19649b802#diff-1444361579758fc6be4458b5105b2ee3ee623d72154094c3fa1fd9056845903eR154

Test summary

```
//docs:update_test                                              (cached) PASSED in 0.0s
//examples/custom_outs:tests_0_test                             (cached) PASSED in 0.1s
//examples/custom_outs:tests_1_test                             (cached) PASSED in 0.0s
//examples/macro:test_test                                      (cached) PASSED in 0.0s
//examples/opaque_src:test_test                                 (cached) PASSED in 0.1s
//examples/out_dir:check_outputs                                (cached) PASSED in 0.1s
//examples/paths:test_test                                      (cached) PASSED in 0.1s
//examples/rc/src:tests_0_test                                  (cached) PASSED in 0.1s
//examples/rc/src:tests_1_test                                  (cached) PASSED in 0.0s
//examples/root_dir:check_outputs                               (cached) PASSED in 0.1s
//examples/simple:test_test                                     (cached) PASSED in 0.1s
//examples/transitive:assertion_test                            (cached) PASSED in 0.0s
//swc/tests:versions_test_test_0                                (cached) PASSED in 0.1s
//examples/directory:minify_test                                         FAILED in 0.0s
//examples/directory:out_dir_test                                        FAILED in 0.0s
//examples/filegroup:check_outputs                                       FAILED in 0.0s
```

merry christmas

**Comment by realtimetodie** [comment]
Fixed an upstream bug, see https://github.com/aspect-build/rules_swc/issues/88

**Comment by realtimetodie** [comment]
The tests are now passing 🎉

**Comment by realtimetodie** [comment]
https://github.com/aspect-build/rules_swc/issues/61

**Comment by alexeagle** [comment]
hey @realtimetodie this is fantastic, I'm watching those upstream PRs.
In theory we could use rules_rust to allow a from-source swc toolchain here, then users could point to an unreleased branch of swc (or they could vendor swc sources into their monorepo so they can make local modifications, though that's essentially a fork)
That's probably not worth it though, the upstream maintainers are active and responsive so I bet you'll be able to get an official swc release that works 100% with Bazel.

**Comment by realtimetodie** [comment]
Yes, the swc team is very responsive. The changes for symbolic links were merged upstream.

**Comment by realtimetodie** [comment]
I updated the [mirror_release.sh](https://github.com/realtimetodie/rules_swc/blob/swc_rust/scripts/mirror_release.sh) script. The binaries are now downloaded from GitHub releases instead from npm

Next, I will refactor repositories.bzl and do the same there.

There is one remaining PR to make the cli work in non tty environments. This is blocking. I'm also working to accept a dictionary to dynamically generate the swrc configuration file, just like we do with tsconfig.json.

**Comment by realtimetodie** [comment]
- I refactored repositories.bzl. The binaries are now downloaded from GitHub releases instead from npm.

- Node.js: All remaining rules_nodejs and npm artifacts were removed. Now, Node.js is only required to run some of the examples.

- In addition, the rule accepts a dictionary to dynamically generate the swrc configuration file. 

    I added tests to validate that the generated configuration works

    https://github.com/realtimetodie/rules_swc/tree/swc_rust/examples/rcdict

- While doing all this, I oriented myself on the existing aspect-build rules and followed the general coding style.

- Upgrade: The `swc` macro is unchanged 🎉 and users do need to upgrade their targets.

As this is a breaking change, I used the opportunity to do some house cleaning (please bear with me).
I renamed the underlying rule from `swc_transpile` -> `swc_compile`. This matches the wording of the SWC project (I think the word "transpile" was originally inspired by the Node-based TypeScript compiler, but this is different in the SWC world). 

Checklist

- [x]  just mirror the rust CLI, no npm packages
- [x]  allow integrity_hashes={"darwin-arm64": "sha512-DuB... to be passed along with version number in WORKSPACE so users don't have to rely on our mirroring
- [x]  look for .swcrc in the package and use it if present, like we do for tsconfig.json in ts_project
- [x]  maybe the swcrc file becomes required to make the rust CLI work. Accept a dictionary to dynamically generate it, just like we do with tsconfig.json
- [x]  get all the tests passing again, and document the breaking changes that were required

Breaking changes will be documented in the final PR.

Blocked by https://github.com/swc-project/swc/pull/6714

**Comment by alexeagle** [comment]
@realtimetodie SWC just cut a release that includes your last fix, so I think it's time to merge this!

**Comment by alexeagle** [comment]
@realtimetodie do you mind signing that CLA so I can include your commits without any fear of licensing issues?

**Comment by realtimetodie** [comment]
signed

**Comment by realtimetodie** [comment]
wow nice, the build is green let's goo

**Comment by alexeagle** [comment]
I'm doing another review/cleanup pass on it. The `k8-fastbuild` is going to be a portability problem
https://github.com/aspect-build/rules_swc/pull/57/files#diff-d1df27564968f40aec96caf8b7e3f176263984b5c8f12dc418a9ea61a19f5073R18

File: swc/private/swc.bzl
Comment (alexeagle) [inline swc/private/swc.bzl:145]: @realtimetodie I think we're missing a test here for the directory output case that an swcrc can be used?

**Comment by realtimetodie** [comment]
The solution to obtain the bazel runfiles directory as an environment variable is very elegant, I like it

File: examples/custom_outs/expected.cjs
Comment (alexeagle) [inline examples/custom_outs/expected.cjs:1]: this change seems incorrect, we output esm modules even though we ran `swc compile --config module.type=commonjs`

Comment (alexeagle) [reply examples/custom_outs/expected.cjs:1]: ah, I see that you're working on it: https://github.com/swc-project/swc/issues/4017#issuecomment-1374141572

I'd rather not hold this release for that, since I've got it all loaded in my brain today. Maybe we just change this example to a TODO.

File: examples/custom_outs/BUILD.bazel
Comment (gregmagolan) [inline examples/custom_outs/BUILD.bazel:35]: name the target `"compile_" + format` to match the new rule name?

File: examples/directory/BUILD.bazel
Comment (gregmagolan) [inline examples/directory/BUILD.bazel:34]: BAZEL_BINDIR? 🤔

File: examples/directory/BUILD.bazel
Comment (gregmagolan) [inline examples/directory/BUILD.bazel:56]: and here too?

File: examples/directory/BUILD.bazel
Comment (alexeagle) [reply examples/directory/BUILD.bazel:34]: just coincidence, not used by rules_js actions - gives an easier way for the bash program to locate files.

Comment (gregmagolan) [reply examples/directory/BUILD.bazel:34]: oh I see
```
readonly in_folder=$(rlocation $TEST_WORKSPACE/$(dirname $TEST_BINARY)/$2/$BAZEL_BINDIR)
```

Comment (alexeagle) [reply examples/directory/BUILD.bazel:34]: maybe it should be using a runfiles lookup instead, I just didn't try very hard :)

File: swc/defs.bzl
Comment (gregmagolan) [inline swc/defs.bzl:71]: so simple :)

File: swc/defs.bzl
Comment (gregmagolan) [inline swc/defs.bzl:70]: what about `%s_swcrc.json` so that in the output tree it is sorted by the target name?

File: swc/private/BUILD.bazel
Comment (gregmagolan) [inline swc/private/BUILD.bazel:20]: 😄

**Review by gregmagolan — APPROVED** [review]
💚

File: swc/defs.bzl
Comment (alexeagle) [reply swc/defs.bzl:71]: yeah thanks @realtimetodie for this line

**Comment by alexeagle** [comment]
> look for .swcrc in the package and use it if present, like we do for tsconfig.json in ts_project

This doesn't actually seem to work, if I drop the `swcrc = ".swcrc"` attribute from an `swc_compile` then it stops using the config file. I'll fix in a follow-up, doesn't really have to be in-scope here as it wasn't a feature before either.

**Pull request merged**
